### PR TITLE
[#159] Add new metric pgagroal_wait_time

### PR DESCRIPTION
**Pull request by An-DJ**
# Issue
#159 

# Task
Add pgagroal_wait_time which characterizes the previous client waiting time.

### CODE REVIEW

**Comment by jesperpedersen** [comment]
I think we can go with second resolution here, and hence use `start_time` as the starting time stamp under `atomic_ulong`.

File: contrib/grafana/dashboard.json [hunk 1]
```diff
@@ -496,6 +496,85 @@
       "title": "Authentication",
       "type": "timeseries"
     },
+    {
+      "datasource": null,
+      "description": "The Waiting time of previous client",
```
Comment (jesperpedersen) [inline contrib/grafana/dashboard.json:501]: "The waiting time for clients"

File: src/include/pgagroal.h [hunk 2]
```diff
@@ -249,6 +249,7 @@ struct prometheus
 
    atomic_ulong client_wait;            /**< The number of waiting clients */
    atomic_ulong client_active;          /**< The number of active clients */
+   atomic_ullong client_wait_time;       /**< The time the client waits */
```
Comment (jesperpedersen) [inline src/include/pgagroal.h:252]: Spacing

File: src/libpgagroal/pool.c [hunk 3]
```diff
@@ -70,16 +71,26 @@ pgagroal_get_connection(char* username, char* database, bool reuse, bool transac
    time_t start_time;
    int best_rule;
    int retries;
+   long long begin_t;
+   struct timeb temp_t;
+
    struct configuration* config;
 
+   struct prometheus* prometheus;
```
Comment (jesperpedersen) [inline src/libpgagroal/pool.c:79]: Extra line feed

File: src/libpgagroal/prometheus.c [hunk 4]
```diff
@@ -687,6 +689,9 @@ home_page(int client_fd)
    data = append(data, "  <h2>pgagroal_failed_servers</h2>\n");
    data = append(data, "  The number of failed servers. Only set if failover is enabled\n");
    data = append(data, "  <p>\n");
+   data = append(data, "  <h2>pgagroal_wait_time</h2>\n");
+   data = append(data, "  The Waiting time of previous client\n");
```
Comment (jesperpedersen) [inline src/libpgagroal/prometheus.c:693]: "The waiting time for clients"

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

**Comment by An-DJ** [comment]
> I think we can go with second resolution here, and hence use `start_time` as the starting time stamp under `atomic_ulong`.

@jesperpedersen Done. I agree with you : )

File: src/libpgagroal/prometheus.c [hunk 5]
```diff
@@ -687,6 +688,9 @@ home_page(int client_fd)
    data = append(data, "  <h2>pgagroal_failed_servers</h2>\n");
    data = append(data, "  The number of failed servers. Only set if failover is enabled\n");
    data = append(data, "  <p>\n");
+   data = append(data, "  <h2>pgagroal_wait_time</h2>\n");
+   data = append(data, "  The waiting time of previous clients\n");
```
Comment (jesperpedersen) [inline src/libpgagroal/prometheus.c:692]: Just put "The waiting time of clients" here

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgagroal/prometheus.c (hunk 5 above)
Comment (An-DJ) [reply src/libpgagroal/prometheus.c]: > Just put "The waiting time of clients" here

Done.

File: contrib/grafana/dashboard.json [hunk 6]
```diff
@@ -496,6 +496,85 @@
       "title": "Authentication",
       "type": "timeseries"
     },
+    {
+      "datasource": null,
+      "description": "The waiting time of previous clients",
```
Comment (jesperpedersen) [inline contrib/grafana/dashboard.json:501]: You missed this one

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: contrib/grafana/dashboard.json (hunk 6 above)
Comment (An-DJ) [reply contrib/grafana/dashboard.json]: > You missed this one

Done.

**Comment by jesperpedersen** [comment]
Thanks for your contribution !

**Pull request merged**